myia/infer/infnode.py

```diff
@@ -89,14 +92,70 @@ def _infer(node, args, unif):
         inp_types = []
         for inp in args:
             inp_types.append((yield Require(inp)))
-        assert len(inp_types) == len(arg_types)
+        assert len(inp_types) == len(arg_types), (
+            f"wrong number of arguments, "
+            f"expected {len(arg_types)}, got {len(inp_types)}"
+        )
         for inp_type, expected_type in zip(inp_types, arg_types):
             autils.unify(expected_type, inp_type, U=unif)
         return autils.reify(return_type, unif=unif.canon)
 
     return inference_function(_infer)
 
 
+class InferenceDefinition(metaclass=OvldMC):
+    """Helper class to represent a signature."""
+
+    __slots__ = "arg_types", "ret_type"
+
+    def __init__(self, *arg_types, ret_type):
+        self.arg_types = tuple(
+            _type_to_abstract(arg_type) for arg_type in arg_types
+        )
+        self.ret_type = _type_to_abstract(ret_type)
+
+    @classmethod
+    def without_value(cls, abstract: data.AbstractAtom):
+        """Get pure type from given abstract, without associated value.
+
+        Useful to lookup node abstract in registered signatures.
+        """
+        return data.AbstractAtom({"interface": abstract.tracks.interface})
```

Comment:
You should be able to do this more generally with something like this:

```python
@abstract_map.variant
def without_value(x: ValueTrack):
    return ValueTrack(ANYTHING)
```

That way, it also works recursively.

Reply:
In fact, it's useless in current branch, so I removed it. I may apply your suggestions in next extracted PRs.